src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -2532,160 +2653,246 @@ var projectUniqueNamesForBuildIntToUpdate
             {
-                await MSBuildRestoreUtility.ReplayWarningsAndErrorsAsync(restoreResult.Result.LockFile?.LogMessages, logger);
-            }
+                var nuGetProjectActions = nugetProjectActionsLookup[buildIntegratedProject.MSBuildProjectPath];
+                var nuGetProjectActionsList = nuGetProjectActions.ToList();
+                var updatedPackageSpec = updatedNugetPackageSpecLookup[buildIntegratedProject.MSBuildProjectPath];
+                var originalPackageSpec = originalNugetPackageSpecLookup[buildIntegratedProject.MSBuildProjectPath];
+                var originalLockFile = lockFileLookup[buildIntegratedProject.MSBuildProjectPath];
+                var sources = nuGetProjectSourceLookup[buildIntegratedProject.MSBuildProjectPath];
 
-            // Build the installation context
-            var originalFrameworks = updatedPackageSpec
-                .TargetFrameworks
-                .ToDictionary(x => x.FrameworkName, x => x.TargetAlias);
+                var allFrameworks = updatedPackageSpec
+                    .TargetFrameworks
+                    .Select(t => t.FrameworkName)
+                    .Distinct()
+                    .ToList();
 
-            var installationContext = new BuildIntegratedInstallationContext(
-                successfulFrameworks,
-                unsuccessfulFrameworks,
-                originalFrameworks);
+                var restoreResult = restoreResults.Single(r =>
+                    string.Equals(
+                        r.SummaryRequest.Request.Project.RestoreMetadata.ProjectPath,
+                        buildIntegratedProject.MSBuildProjectPath,
+                        StringComparison.OrdinalIgnoreCase));
+
+                var unsuccessfulFrameworks = restoreResult
+                    .Result
+                    .CompatibilityCheckResults
+                    .Where(t => !t.Success)
+                    .Select(t => t.Graph.Framework)
+                    .Distinct()
+                    .ToList();
+
+                var successfulFrameworks = allFrameworks
+                    .Except(unsuccessfulFrameworks)
+                    .ToList();
+
+                var firstAction = nuGetProjectActionsList[0];
+
+                // If the restore failed and this was a single package install, try to install the package to a subset of
+                // the target frameworks.
+                if (nuGetProjectActionsList.Count == 1 &&
+                    firstAction.NuGetProjectActionType == NuGetProjectActionType.Install &&
+                    successfulFrameworks.Any() &&
+                    unsuccessfulFrameworks.Any() &&
+                    !restoreResult.Result.Success &&
+                    // Exclude upgrades, for now we take the simplest case.
+                    !PackageSpecOperations.HasPackage(originalPackageSpec, firstAction.PackageIdentity.Id))
+                {
+                    updatedPackageSpec = originalPackageSpec.Clone();
```

Comment:
Why clone again? Cannot be used the https://github.com/NuGet/NuGet.Client/blob/24c713cd927fe4920ec5ecc6fe98033f84eab727/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2780